doc/user-guide/testing.rst

```diff
@@ -0,0 +1,304 @@
+
+.. ipython:: python
+
+    # Generates a Variable with no attributes
+    xrst.variables(attrs=st.just({})).example()
+
+Through a combination of chaining strategies and fixing arguments, you can specify quite complicated requirements on the
+objects your chained strategy will generate.
+
+.. ipython:: python
+
+    fixed_x_variable_y_maybe_z = st.fixed_dictionaries(
+        {"x": st.just(2), "y": st.integers(3, 4)}, optional={"z": st.just(2)}
+    )
+    fixed_x_variable_y_maybe_z.example()
+
+    special_variables = xrst.variables(dims=fixed_x_variable_y_maybe_z)
+
+    special_variables.example()
+    special_variables.example()
+
+Here we have used one of hypothesis' built-in strategies :py:func:`hypothesis.strategies.fixed_dictionaries` to create a
+strategy which generates mappings of dimension names to lengths (i.e. the ``size`` of the xarray object we want).
+This particular strategy will always generate an ``x`` dimension of length 2, and a ``y`` dimension of
+length either 3 or 4, and will sometimes also generate a ``z`` dimension of length 2.
+By feeding this strategy for dictionaries into the ``dims`` argument of xarray's :py:func:`~st.variables` strategy,
+we can generate arbitrary :py:class:`~xarray.Variable` objects whose dimensions will always match these specifications.
+
+Generating Duck-type Arrays
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Xarray objects don't have to wrap numpy arrays, in fact they can wrap any array type which presents the same API as a
+numpy array (so-called "duck array wrapping", see :ref:`wrapping numpy-like arrays <internals.duckarrays>`).
+
+Imagine we want to write a strategy which generates arbitrary ``Variable`` objects, each of which wraps a
+:py:class:`sparse.COO` array instead of a ``numpy.ndarray``. How could we do that? There are two ways:
+
+1. Create a xarray object with numpy data and use the hypothesis' ``.map()`` method to convert the underlying array to a
+different type:
+
+.. ipython:: python
+
+    import sparse
+
+.. ipython:: python
+
+    def convert_to_sparse(var):
+        var.data = sparse.COO.from_numpy(var.to_numpy())
+        return var
+
+.. ipython:: python
+
+    sparse_variables = xrst.variables(dims=xrst.dimension_names(min_dims=1)).map(
+        convert_to_sparse
+    )
+
+    sparse_variables.example()
+    sparse_variables.example()
+
+2. Pass a function which returns a strategy which generates the duck-typed arrays directly to the ``array_strategy_fn`` argument of the xarray strategies:
+
+.. ipython:: python
+
+    @st.composite
+    def sparse_random_arrays(draw, shape: tuple[int]) -> sparse._coo.core.COO:
+        """Strategy which generates random sparse.COO arrays"""
+        if shape is None:
+            shape = draw(npst.array_shapes())
+        density = draw(st.integers(min_value=0, max_value=1))
+        return sparse.random(
+            shape=shape, density=density
+        )  # note sparse.random does not accept a dtype kwarg
```

Comment:
if I understand correctly, this can be rewritten to use `st.builds`:
```suggestion
    def sparse_random_arrays(shape: tuple[int]) -> sparse._coo.core.COO:
        """Strategy which generates random sparse.COO arrays"""
        if shape is None:
            shape = npst.array_shapes()
        else:
            shape = st.just(shape)
        density = st.integers(min_value=0, max_value=1)
        # note sparse.random does not accept a dtype kwarg
        return st.builds(sparse.random, shape=shape, density=density)
```